File: frontend/src/interfaces/Schedule3Response.ts
// Mirrors the backend Schedule3Response DTO (Stories 4.1/4.2), the AD-12 aggregate document. Numbers
// are nullable; Jackson omits nulls, so an absent line/block member simply won't be in the JSON. Every
// crown value, timber cost/perUnit, subtotal, total and Total Overhead figure is computed server-side
// and is read-only here — never sent on a write, and the server is the sole authority for every stored
// figure.
//
// They ARE mirrored for display while the schedule is being edited, so the read-only cells track entry
// before Save the way legacy did (defect #291; spine AD-5 amended 2026-08-20). That mirror lives in
// `components/schedule3/derived.ts` and nowhere else, and the Save echo supersedes it. Note the
// Scaling (33) line's `pop` is itself derived from the two timber volumes, and that Schedule 3 rounds
// $/m³ with the LEGACY scale-2 rule (divide at scale 10, then scale 2), not Schedules 2/4's scale 4.

import type { MessageInfo } from '@/interfaces/Schedule1Response'

// One fixed admin-cost line in the three-column model. `crown` is derived (harvest − pop), read-only.
export interface CostLine {
  costItemCode: number
  harvest: number | null
  pop: number | null
  crown: number | null
}

// A timber / overhead block: PO&P Timber, Crown Timber, Total Overhead. `volume` is entered (except
// Total Overhead's, which is derived); `cost` and `perUnit` are server-computed read-only.
export interface TimberBlock {
  volume: number | null
  cost: number | null
  perUnit: number | null
}

// A derived three-column money total (Subtotal Other Costs, Subtotal Actual Costs, Included
// Unacceptable Costs, Total Costs). All fields server-computed read-only.
export interface ThreeColumnTotal {
  harvest: number | null
  pop: number | null
  crown: number | null
}

export default interface Schedule3Response {
  millId: number
  year: number
  trackStatus: string | null
  editable: boolean
  // Optional AND nullable (defect #292): null until the schedule is saved, and the app-wide Jackson
  // `default-property-inclusion: non_null` then omits the key — so readers must use a loose `!= null`.
  // Unreachable through this page today (the GET 404s when unsaved), typed honestly so it stays safe.
  revisionCount?: number | null
  // Override Harvest ⁄ Total PO&P indicator ("Y"/"N"), from the summary LOCATION column; defaults "N".
  overrideHarvestTotalPop: string | null
  comments: string | null
  // The 11 fixed admin-cost lines (codes 27–37), each with harvest/pop/derived crown.
  lineItems: CostLine[]
  popTimber: TimberBlock
  crownTimber: TimberBlock
  totalOverhead: TimberBlock
  subtotalOtherCosts: ThreeColumnTotal
  subtotalActualCosts: ThreeColumnTotal
  includedUnacceptableCosts: ThreeColumnTotal
  totalCosts: ThreeColumnTotal
  // Sub-page row-group counts (CNT-001) shown on the two sub-page links (Story 4.4 owns the sub-pages).
  otherAcceptableCount: number
  unacceptableCount: number
  // Advisory, non-blocking messages on a mutation echo (BR-09 crown-push WRN-001/002 ride here). Story 4.2.
  warnings?: MessageInfo[]
  // Success message on a mutation echo (AD-8), else null/absent on the GET.
  message?: MessageInfo | null
}

File: backend/src/main/java/ca/bc/gov/nrs/ilcr/schedule3/dto/Schedule3Response.java
package ca.bc.gov.nrs.ilcr.schedule3.dto;

import java.util.List;

/**
 * The Schedule 3 aggregate document (AD-5, AD-12) — the pinned GET response for the Forest Management
 * Administration Costs report. All derived/read-only fields (each line's {@code crown}, the timber
 * costs/{@code perUnit}, the four total blocks, {@code trackStatus}, {@code editable}, the counts)
 * are computed server-side and are ignored/rejected on write (Story 4.2). {@code revisionCount} is the
 * optimistic-lock token echoed for the Story 4.2 PUT. {@code message} is null on GET (Jackson omits
 * it) and carries the success message on the mutation echo (AD-8/EQ-M3).
 *
 * @param millId the mill id
 * @param year the reporting year
 * @param trackStatus the Schedules 1–10 track status ({@code ILCR_MILL_REPORT_STATUS_CODE}, AD-9)
 * @param editable server-authoritative: {@code EDIT_SCHEDULE} held AND track is Draft
 * @param revisionCount optimistic-lock token (summary {@code REVISION_COUNT})
 * @param overrideHarvestTotalPop the Override Harvest/Total PO&amp;P indicator ("Y"/"N"), from the
 *     summary {@code LOCATION} column (legacy carrier); defaults to "N" (BR-10, Story 4.2 writes it)
 * @param comments the free-text comments (max 3500 chars)
 * @param lineItems the 11 fixed admin-cost lines (27–37), each with harvest/pop/derived crown
 * @param popTimber the PO&amp;P Timber block (item 118 volume; cost/perUnit derived)
 * @param crownTimber the Crown Timber block (item 119 volume; drives the BR-09 push in Story 4.2)
 * @param totalOverhead the Total Overhead block (all derived)
 * @param subtotalOtherCosts the Other Acceptable Costs subtotal (from item-124 rows)
 * @param subtotalActualCosts the Subtotal Actual Costs (sum of the fixed lines + Subtotal Other Costs)
 * @param includedUnacceptableCosts the Included Unacceptable Costs (item-38 rows + Annual Rents harvest)
 * @param totalCosts the Total Costs (Subtotal Actual &minus; Included Unacceptable)
 * @param otherAcceptableCount the Other Acceptable Costs sub-page row-group count (CNT-001)
 * @param unacceptableCount the Included Unacceptable Costs sub-page row count (CNT-001; +1 for Annual Rents)
 * @param warnings advisory, non-blocking messages carried on a mutation echo (empty on GET); the BR-09
 *     Crown Timber push outcome (WRN-001 applied / WRN-002 not-opened) rides here on the Story 4.2 PUT
 * @param message the success message on a mutation echo, else null
 */
public record Schedule3Response(
    long millId,
    int year,
    String trackStatus,
    boolean editable,
    Integer revisionCount,
    String overrideHarvestTotalPop,
    String comments,
    List<CostLine> lineItems,
    TimberBlock popTimber,
    TimberBlock crownTimber,
    TimberBlock totalOverhead,
    ThreeColumnTotal subtotalOtherCosts,
    ThreeColumnTotal subtotalActualCosts,
    ThreeColumnTotal includedUnacceptableCosts,
    ThreeColumnTotal totalCosts,
    int otherAcceptableCount,
    int unacceptableCount,
    List<MessageInfo> warnings,
    MessageInfo message) {

  /** A copy of this document carrying the given success message (for the PUT echo, AD-8). */
  public Schedule3Response withMessage(MessageInfo message) {
    return new Schedule3Response(
        millId, year, trackStatus, editable, revisionCount, overrideHarvestTotalPop, comments,
        lineItems, popTimber, crownTimber, totalOverhead, subtotalOtherCosts, subtotalActualCosts,
        includedUnacceptableCosts, totalCosts, otherAcceptableCount, unacceptableCount, warnings,
        message);
  }

  /** A copy of this document carrying the given advisory warnings (e.g. the BR-09 crown-push outcome). */
  public Schedule3Response withWarnings(List<MessageInfo> warnings) {
    return new Schedule3Response(
        millId, year, trackStatus, editable, revisionCount, overrideHarvestTotalPop, comments,
        lineItems, popTimber, crownTimber, totalOverhead, subtotalOtherCosts, subtotalActualCosts,
        includedUnacceptableCosts, totalCosts, otherAcceptableCount, unacceptableCount, warnings,
        message);
  }
}
